src/main/java/org/dbos/apiary/mysql/MysqlConnection.java

```diff
@@ -0,0 +1,188 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+public class MysqlConnection implements ApiarySecondaryConnection {
+    private static final Logger logger = LoggerFactory.getLogger(MysqlConnection.class);
+
+    private final MysqlDataSource ds;
+    private final ThreadLocal<Connection> connection;
+
+    private final Map<String, Map<String, Set<Long>>> committedWrites = new ConcurrentHashMap<>();
+    private final Lock validationLock = new ReentrantLock();
+
+    public MysqlConnection(String hostname, Integer port, String databaseName, String databaseUsername, String databasePassword) throws SQLException {
+        this.ds = new MysqlDataSource();
+        // Set dataSource Properties
+        this.ds.setServerName(hostname);
+        this.ds.setPortNumber(port);
+        this.ds.setDatabaseName(databaseName);
+        this.ds.setUser(databaseUsername);
+        this.ds.setPassword(databasePassword);
+
+        this.connection = ThreadLocal.withInitial(() -> {
+            try {
+                Connection conn = ds.getConnection();
+                conn.setAutoCommit(true);
+                // conn.setTransactionIsolation(Connection.TRANSACTION_REPEATABLE_READ);
+                return conn;
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+            return null;
+        });
+
+        try {
+            Connection testConn = ds.getConnection();
+            testConn.close();
+        } catch (SQLException e) {
+            logger.info("Failed to connect to MySQL");
+            throw new RuntimeException("Failed to connect to MySQL");
+        }
+    }
+
+    public void dropTable(String tableName) throws SQLException {
+        Connection conn = ds.getConnection();
+        Statement truncateTable = conn.createStatement();
+        truncateTable.execute(String.format("DROP TABLE IF EXISTS %s;", tableName));
+        truncateTable.close();
+        conn.close();
+    }
+
+    public void createTable(String tableName, String specStr) throws SQLException {
+        // TODO: How do we interact with the original primary key columns to avoid conflicts? Add Apiary columns as part of primary key? For now, assume no primary key columns.
+        // Automatically add three additional columns: apiaryID, beginVersion, endVersion.
+        Connection conn = ds.getConnection();
+        Statement s = conn.createStatement();
+        String apiaryTable = String.format(
+                "CREATE TABLE IF NOT EXISTS %s (%s VARCHAR(256) NOT NULL, %s BIGINT, %s BIGINT, %s);"
+        , tableName, MysqlContext.apiaryID, MysqlContext.beginVersion, MysqlContext.endVersion, specStr);
```

Comment:
We might later need indexes on these added columns to speed up the queries. Otherwise, it might turn into scans? Not sure.

Reply:
We can optimize them later.

Reply:
Yeah, agree. For now, I'm not sure which and how many indices we need. We need more benchmarking and tuning in the future PRs.